### PR TITLE
Add an 'ignore matches' predicate

### PR DESCRIPTION
**Pull request by jonathonherbert**
## What does this change?

Adds an 'ignore matches' predicate that the plugin consumer can supply. This predicate receives every match supplied to the plugin on init or as a result of a successful response. If it returns false, the match is not added to the plugin state.

An example use case at the Guardian: in our text editor, our users have the ability to mark text ranges as 'correct'. We don't want to see matches for these ranges.

## How to test

The added tests should give us an idea of how this works.

You could also spin this up in a test environment and add a predicate to check it behaves as expected.

## How can we measure success?

Matches that ... match the predicate are not included in the plugin state.

### CODE REVIEW

**Comment by SHession** [comment]
Would it be possible to get this rebased for a more accurate DIFF?

**Comment by jonathonherbert** [comment]
@SHession great point, I've just done that 👍

File: src/ts/state/reducer.ts [hunk 1]
```diff
@@ -146,25 +154,29 @@ export const PROSEMIRROR_TYPERIGHTER_ACTION = "PROSEMIRROR_TYPERIGHTER_ACTION";
...
+    dirtiedRanges: [],
+    currentMatches: [],
+    selectedMatch: undefined,
+    hoverId: undefined,
+    hoverInfo: undefined,
+    requestsInFlight: {},
+    requestPending: false,
+    requestErrors: []
+  };
+  return addMatchesToState(initialState, doc, matches, ignoreMatch);
```
Comment (jonathonherbert) [inline src/ts/state/reducer.ts:176]: The consumer can add matches when the plugin initialises – we need to run those matches through the `ignoreMatch` predicate, so a new helper function, `addMatchesToState`, takes care of that task.

File: src/ts/state/test/reducer.spec.ts [hunk 2]
```diff
@@ -22,37 +22,63 @@ import {
...
+import { getBlocksFromDocument } from "../../utils/prosemirror";
 
-const reducer = createReducer(expandRangesToParentBlockNode);
+const reducer = createReducer(expandRangesToParentBlockNode, () => true);
+
+/**
+ * Create a plugin state, creating the given matches and
+ * their decorations from the given spec.
+ */
+const createStateWithMatches = (
```
Comment (jonathonherbert) [inline src/ts/state/test/reducer.spec.ts:46]: There's a bit of scaffolding necessary to test the ignoreMatches predicate, which I've abstracted here. This is likely to be useful for other tests, but I haven't included that change here to keep the diff manageable.

File: src/ts/test/helpers/fixtures.ts [hunk 3]
```diff
@@ -50,42 +50,69 @@ export const createBlock = (
...
-  blocks: [
-    {
-      id: createBlockId(0, from, to),
-      from,
-      to,
-      text: "block text"
-    }
-  ],
-  matches: [
+  specs: ICreateMatcherResponseSpec[],
```
Comment (jonathonherbert) [inline src/ts/test/helpers/fixtures.ts:64]: I had to add a bit of detail to the fixture here to make it play nicely with the plugin state – we're testing more of it now, so we need a more accurate representation of what we'd get from a response.

File: src/ts/utils/decoration.ts [hunk 4]
```diff
@@ -43,10 +43,7 @@ export const createDebugDecorationFromRange = (range: IRange, dirty = true) => {
 export const removeDecorationsFromRanges = (
   decorationSet: DecorationSet,
   ranges: IRange[],
-  types = [
-    DECORATION_MATCH,
-    DECORATION_MATCH_HEIGHT_MARKER
-  ]
+  types = [DECORATION_MATCH, DECORATION_MATCH_HEIGHT_MARKER]
```
Comment (jonathonherbert) [inline src/ts/utils/decoration.ts:46]: Prettier. Very tempted add a lint step to our git hooks.

File: src/ts/createTyperighterPlugin.ts [hunk 5]
```diff
@@ -53,20 +58,21 @@ const createTyperighterPlugin = <TMatch extends IMatch>(
   const {
     expandRanges = expandRangesToParentBlockNode,
     matches = [],
-    isActive
+    isActive,
+    ignoreMatch = () => true
```
Comment (aug24) [inline src/ts/createTyperighterPlugin.ts:62]: Err.  Shouldn't this be 'false'?

File: src/ts/state/helpers.ts [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+import { createDecorationsForMatch } from "../utils/decoration";
+import { DecorationSet } from "prosemirror-view";
+
+export const addMatchesToState = <TMatch extends IMatch>(
+  state: IPluginState<TMatch>,
+  doc: any,
+  matches: TMatch[],
+  ignoreMatch: IIgnoreMatch
+) => {
+  const matchesToApply = matches.filter(ignoreMatch);
```
Comment (aug24) [inline src/ts/state/helpers.ts:12]: And surely this should be filter(!ignoreMatch) ?

File: src/ts/createTyperighterPlugin.ts (hunk 5 above)
Comment (jonathonherbert) [reply src/ts/createTyperighterPlugin.ts]: Yes. Yes it should.

File: src/ts/state/helpers.ts (hunk 6 above)
Comment (jonathonherbert) [reply src/ts/state/helpers.ts]: As above 😅

File: src/ts/createTyperighterPlugin.ts [hunk 7]
```diff
@@ -54,20 +59,21 @@ const createTyperighterPlugin = <TMatch extends IMatch>(
     expandRanges = expandRangesToParentBlockNode,
     matches = [],
     isActive = true,
+    ignoreMatch = () => false
```
Comment (aug24) [inline src/ts/createTyperighterPlugin.ts:62]: `() => false` appears in four places.  Can it be DRYed out and given a name?

Comment (jonathonherbert) [reply src/ts/createTyperighterPlugin.ts:62]: Great point, let's provide a default implementation.

File: src/ts/createTyperighterPlugin.ts [hunk 8]
```diff
@@ -37,6 +37,11 @@ export interface IPluginOptions<TMatch extends IMatch = IMatch> {
    * Is the plugin active as it starts?
    */
   isActive?: boolean;
+
+  /**
+   * Ignore matches when this predicate returns false.
```
Comment (aug24) [inline src/ts/createTyperighterPlugin.ts:42]: IS THIS COMMENT CORRECT?

File: src/ts/createTyperighterPlugin.ts (hunk 7 above)
Comment (aug24) [reply src/ts/createTyperighterPlugin.ts:62]: Did you?

File: src/ts/state/reducer.ts [hunk 9]
```diff
@@ -92,6 +93,14 @@ export interface IBlockInFlight {
   block: IBlock;
 }
 
+/**
+ * A consumer-supplied predicate that allows consumers to ignore matches.
+ * Handy when, for example, consumers know that parts of the document are
+ * exempt from checks.
+ */
+export type IIgnoreMatch = (match: IMatch) => boolean;
+export const defaultIgnoreMatch: IIgnoreMatch = () => false;
```
Comment (aug24) [inline src/ts/state/reducer.ts:102]: You're going to hate me, but I don't like the name.  Why not 'includeAll' or something with business meaning?

Comment (jonathonherbert) [reply src/ts/state/reducer.ts:102]: > Did you?
I'm still blaming the heat 😅 

> You're going to hate me, but I don't like the name. Why not 'includeAll' or something with business meaning?
Naw, I completely agree. Like you say, business meaning > what the thing _is_. Just changed 👍

File: src/ts/createTyperighterPlugin.ts (hunk 8 above)
Comment (jonathonherbert) [reply src/ts/createTyperighterPlugin.ts]: ... now it is 😅

**Pull request merged**